src/cascadia/TerminalControl/TSFInputControl.cpp

```diff
@@ -0,0 +1,380 @@
+
+        // add the Textblock to the Canvas
+        _canvas.Children().Append(_textBlock);
+
+        // set the content of this control to be the Canvas
+        this->Content(_canvas);
+
+        // Create a CoreTextEditingContext for since we are acting like a custom edit control
+        auto manager = Core::CoreTextServicesManager::GetForCurrentView();
+        _editContext = manager.CreateEditContext();
+
+        // sets the Input Pane display policy to Manual for now so that it can manually show the
+        // software keyboard when the control gains focus and dismiss it when the control loses focus.
+        // TODO GitHub #3639: Should Input Pane display policy be Automatic
+        _editContext.InputPaneDisplayPolicy(Core::CoreTextInputPaneDisplayPolicy::Manual);
+
+        // set the input scope to Text because this control is for any text.
+        _editContext.InputScope(Core::CoreTextInputScope::Text);
+
+        _editContext.TextRequested({ this, &TSFInputControl::_textRequestedHandler });
+
+        _editContext.SelectionRequested({ this, &TSFInputControl::_selectionRequestedHandler });
+
+        _editContext.FocusRemoved({ this, &TSFInputControl::_focusRemovedHandler });
+
+        _editContext.TextUpdating({ this, &TSFInputControl::_textUpdatingHandler });
+
+        _editContext.SelectionUpdating({ this, &TSFInputControl::_selectionUpdatingHandler });
+
+        _editContext.FormatUpdating({ this, &TSFInputControl::_formatUpdatingHandler });
+
+        _editContext.LayoutRequested({ this, &TSFInputControl::_layoutRequestedHandler });
+
+        _editContext.CompositionStarted({ this, &TSFInputControl::_compositionStartedHandler });
+
+        _editContext.CompositionCompleted({ this, &TSFInputControl::_compositionCompletedHandler });
+    }
+
+    // Method Description:
+    // - NotifyFocusEnter handler for notifying CoreEditTextContext of focus enter
+    //   when TerminalControl receives focus.
+    // Arguments:
+    // - <none>
+    // Return Value:
+    // - <none>
+    void TSFInputControl::NotifyFocusEnter()
+    {
+        if (_editContext != nullptr)
+        {
+            _editContext.NotifyFocusEnter();
+        }
+    }
+
+    // Method Description:
+    // - NotifyFocusEnter handler for notifying CoreEditTextContext of focus leaving.
+    //   when TerminalControl no longer has focus.
+    // Arguments:
+    // - <none>
+    // Return Value:
+    // - <none>
+    void TSFInputControl::NotifyFocusLeave()
+    {
+        if (_editContext != nullptr)
+        {
+            // _editContext.NotifyFocusLeave(); TODO GitHub #3645: Enabling causes IME to no longer show up, need to determine if required
+        }
+    }
+
+    // Method Description:
+    // - Scales a Rect based on a scale factor
+    // Arguments:
+    // - rect: Rect to scale by scale
+    // - scale: amount to scale rect by
+    // Return Value:
+    // - Rect scaled by scale
+    inline Rect ScaleRect(Rect rect, double scale)

```

Comment:
This looks like another good `Utils` function

Reply:
Moved to Utils will be in next push